sky/data/skystorage.py

```diff
@@ -0,0 +1,227 @@
+def csync(src: str, storetype: str, bucketname: str, num_threads: int,
+          interval: int, delete: bool, lock: bool, no_follow_symlinks: bool):
+    """
+    Syncs the source to the bucket every INTERVAL seconds. Creates a lock file
+    while sync command is runninng and removes it when completed.
+    """
+    base_dir = os.path.expanduser(C_SYNC_FILE_PATH)
+    os.makedirs(base_dir, exist_ok=True)
+    lock_file_name = f'csync_{storetype}_{bucketname}.lock'
+    lock_path = os.path.expanduser(os.path.join(base_dir, lock_file_name))
+
+    # When the csync daemon was previously terminated abnormally,
+    # it may still be holding the lock. This lock will be reset after removal.
+    if os.path.exists(lock_path):
+        os.remove(lock_path)
+
+    while True:
+        with contextlib.ExitStack() as stack:
+            if lock:
+                stack.enter_context(filelock.FileLock(lock_path))
+            start_time = time.time()
+            # TODO: add try-except block
+            run_sync(src, storetype, bucketname, num_threads, interval, delete,
+                     no_follow_symlinks)
+            end_time = time.time()
+        # the time took to sync gets reflected to the INTERVAL
+        elapsed_time = int(end_time - start_time)
+        updated_interval = update_interval(interval, elapsed_time)
+        if lock and os.path.exists(lock_path):
+            os.remove(lock_path)
+        time.sleep(updated_interval)
+
+
+@main.command()
+def terminate() -> None:
+    """
+    Terminates all the CSYNC daemon running after checking if all the
+    sync process has completed.
+    """
+    # TODO: Currently, this terminates all the CSYNC daemon by default.
+    # Make an option of --all to terminate all and make the default
+    # behavior to take a source name to terminate only one daemon.
+    # Call the function to terminate the csync processes here
+    process_dict: Dict[int, Tuple[str, str]] = {}
+
+    # check all the running processes to see if CSYNC daemon is running
+    for proc in psutil.process_iter(['cmdline']):
+        cmd = proc.info['cmdline']
+        if 'csync' in cmd:
+            # cmd[5] is store type and cmd[6] is the bucket name
+            process_dict[proc.pid] = (cmd[5], cmd[6])
+
+    if len(process_dict) > 0:
+        while True:
+            running_sync = dict(process_dict)
+            if not running_sync:
+                break
+            for pid in list(running_sync):
+                store_type = running_sync[pid][0]
+                bucket_name = running_sync[pid][1]
+                lock_file_name = f'csync_{store_type}_{bucket_name}.lock'
+                lock_file_path = os.path.expanduser(
+                    os.path.join(C_SYNC_FILE_PATH, lock_file_name))
+                if not os.path.exists(lock_file_path):
+                    # kill csync process
+                    psutil.Process(pid).terminate()
+                    del process_dict[pid]
+            time.sleep(5)
```

Comment:
This seems very hacky. We probably need to do it in a more structured way instead of doing the string processing.

Reply:
@Michaelvll Do you mind hinting on what kind of structured approach I can take? Another approach I can think of was to involve a local db, but was wondering if there is another way as we sort of determined to not set up a local db for this. Or perhaps do you think we need it?

Reply:
Creating a local db to store CSYNC related information to keep this more structured.